Common/Securities/Futures/FutureFilterUniverse.cs

```diff
@@ -0,0 +1,205 @@
+        /// would exclude contracts expiring in less than 10 days</param>
+        /// <param name="maxExpiry">The maxmium time until expiry to include, for example, TimeSpan.FromDays(10)
+        /// would exclude contracts expiring in more than 10 days</param>
+        /// <returns></returns>
+        public FutureFilterUniverse Expiration(TimeSpan minExpiry, TimeSpan maxExpiry)
+        {
+            if (_underlying == null)
+            {
+                return this;
+            }
+
+            if (maxExpiry > Time.MaxTimeSpan) maxExpiry = Time.MaxTimeSpan;
+
+            var minExpiryToDate = _underlying.Time.Date + minExpiry;
+            var maxExpiryToDate = _underlying.Time.Date + maxExpiry;
+
+            // ReSharper disable once PossibleMultipleEnumeration - ReSharper is wrong here due to the ToList call
+            var filtered =
+                   from symbol in _allSymbols
+                   let contract = symbol.ID
+                   where contract.Date >= minExpiryToDate
+                      && contract.Date <= maxExpiryToDate
+                   select symbol;
+
+            _allSymbols = filtered;
+            return this;
+        }
+
+        /// <summary>
+        /// Applies filter selecting futures contracts based on expiration cycles. See <see cref="FutureExpirationCycles"/> for details 
+        /// </summary>
+        /// <param name="months"></param>
+        /// <returns></returns>
+        public FutureFilterUniverse ExpirationCycle(int[] months)
+        {
+            var monthHashSet = months.ToHashSet();
+            return this.Where(x => monthHashSet.Contains(x.ID.Date.Month));
+        }
+
+        /// <summary>
+        /// IEnumerable interface method implementation
+        /// </summary>
+        public IEnumerator<Symbol> GetEnumerator()
+        {
+            return _allSymbols.GetEnumerator();
+        }
+
+        /// <summary>
+        /// IEnumerable interface method implementation
+        /// </summary>
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return _allSymbols.GetEnumerator();
+        }
+    }
+
+    /// <summary>
+    /// Extensions for Linq support
+    /// </summary>
+    public static class FutureFilterUniverseEx
+    {
+        /// <summary>
+        /// Filters universe 
+        /// </summary>
+        /// <param name="universe"></param>
+        /// <param name="predicate"></param>
+        /// <returns></returns>
+        public static FutureFilterUniverse Where(this FutureFilterUniverse universe, Func<Symbol, bool> predicate)
+        {
+            universe._allSymbols = universe._allSymbols.Where(predicate);
+            universe._isDynamic = true;
+            return universe;
+        }
+
+        /// <summary>
+        /// Maps universe 
+        /// </summary>
+        public static FutureFilterUniverse Select(this FutureFilterUniverse universe, Func<Symbol, Symbol> mapFunc)
```

Comment:
Awesome knowledge of the C# compiler. This stuff enables the query syntax, super cool!!